File: bibgraph/ingest_html/mathml.py
"""Convert presentation MathML to KaTeX-renderable LaTeX, via ``pandoc``.

This is a *fallback*: A&A (and most MathJax-rendered sites) expose the original
LaTeX directly (e.g. a ``data-latex`` attribute or a ``<annotation
encoding="application/x-tex">`` inside the MathML ``<semantics>``), which is
always preferred. Only when no source LaTeX is available do we hand the
presentation MathML to pandoc, which reads MathML embedded in HTML natively and
emits clean LaTeX (``\\frac``, ``^``, ``\\begin{array}`` …).
"""

from __future__ import annotations

import logging
import re
import shutil
import subprocess

log = logging.getLogger("bibgraph.html.mathml")

_PANDOC = shutil.which("pandoc")

# pandoc wraps display math in \[..\] and inline in \(..\); strip either.
_DELIMS = (("\\[", "\\]"), ("\\(", "\\)"), ("$$", "$$"), ("$", "$"))


def have_pandoc() -> bool:
    return _PANDOC is not None


def annotation_latex(math_el) -> str | None:
    """Original TeX from a MathML ``<annotation encoding="application/x-tex">``."""
    ann = math_el.find(
        "annotation", attrs={"encoding": ["application/x-tex", "application/x-latex"]}
    )
    if ann is None:
        # some emitters omit/spell the encoding differently
        for a in math_el.find_all("annotation"):
            enc = (a.get("encoding") or "").lower()
            if "tex" in enc:
                ann = a
                break
    if ann is not None:
        tex = ann.get_text().strip()
        return tex or None
    return None


def mathml_to_latex(math_html: str) -> str | None:
    """Convert one ``<math>...</math>`` HTML string to LaTeX (or None)."""
    if not _PANDOC or not math_html or "<math" not in math_html:
        return None
    doc = f"<!DOCTYPE html><html><body><p>{math_html}</p></body></html>"
    try:
        out = subprocess.run(
            [_PANDOC, "-f", "html", "-t", "latex"],
            input=doc, capture_output=True, text=True, timeout=20,
        )
    except Exception as e:                            # pragma: no cover
        log.warning("pandoc invocation failed: %s", e)
        return None
    if out.returncode != 0:
        log.debug("pandoc rc=%s: %s", out.returncode, out.stderr[:200])
        return None
    return _katexify(strip_math_delims(out.stdout.strip()))


# pandoc emits a few LaTeX commands KaTeX doesn't implement; map them to
# KaTeX-supported equivalents (spacing differences are cosmetic).
_MSPACE_RE = re.compile(r"\\mspace\s*\{[^}]*\}")


def _katexify(latex: str) -> str:
    latex = _MSPACE_RE.sub(r"\\;", latex)        # \mspace{6mu} -> \;
    latex = latex.replace("\\medspace", "\\;").replace("\\thickspace", "\\;")
    return latex


def strip_math_delims(latex: str) -> str:
    """Drop one layer of surrounding $/$$/\\[..\\]/\\(..\\) delimiters."""
    s = (latex or "").strip()
    for lo, hi in _DELIMS:
        if s.startswith(lo) and s.endswith(hi) and len(s) > len(lo) + len(hi):
            return s[len(lo):len(s) - len(hi)].strip()
    return s

File: packages/core/src/pipelines/html/mathml.ts
/**
 * The DOM half of `bibgraph/ingest_html/mathml.py`: reading the publisher's
 * original LaTeX out of a MathML `<annotation>` element (pure), plus
 * `strip_math_delims` (now consumed by core's adapters — M2 had parked it in
 * infra because nothing in core used it yet). The pandoc-backed
 * `mathml_to_latex` itself stays an injected port (`HtmlMathmlPort`); infra's
 * `latex/pandoc.ts` implements it (with `katexify` applied to pandoc's output,
 * exactly like the Python).
 */

import type { Element } from "domhandler";
import { findAllWhere, getAttr, getText } from "./dom.js";

// pandoc wraps display math in \[..\] and inline in \(..\); strip either.
const DELIMS: ReadonlyArray<readonly [string, string]> = [
  ["\\[", "\\]"],
  ["\\(", "\\)"],
  ["$$", "$$"],
  ["$", "$"],
];

/** `annotation_latex`: original TeX from `<annotation encoding="application/x-tex">`. */
export function annotationLatex(mathEl: Element): string | undefined {
  let ann = findAllWhere(
    mathEl,
    (e) =>
      e.name === "annotation" &&
      ["application/x-tex", "application/x-latex"].includes(getAttr(e, "encoding") ?? "")
  )[0];
  if (!ann) {
    // some emitters omit/spell the encoding differently
    ann = findAllWhere(
      mathEl,
      (e) => e.name === "annotation" && (getAttr(e, "encoding") ?? "").toLowerCase().includes("tex")
    )[0];
  }
  if (ann) {
    const tex = getText(ann).trim();
    return tex || undefined;
  }
  return undefined;
}

/** `strip_math_delims`: drop one layer of surrounding $/$$/\[..\]/\(..\). */
export function stripMathDelims(latex: string): string {
  const s = (latex || "").trim();
  for (const [lo, hi] of DELIMS) {
    if (s.startsWith(lo) && s.endsWith(hi) && s.length > lo.length + hi.length) {
      return s.slice(lo.length, s.length - hi.length).trim();
    }
  }
  return s;
}
